### PR TITLE
Add isderiving utility

### PR DESCRIPTION
**Pull request by DhairyaLGandhi**
cc @ChrisRackauckas who wanted a utility like this

### CODE REVIEW

**Comment by CarloLucibello** [comment]
I would also consider the name `isdifferentiating`

**Comment by ChrisRackauckas** [comment]
https://github.com/FluxML/Zygote.jl/issues/850 is an example of how this fails.

**Comment by mcabbott** [comment]
Agree this needs a better name. As verbs, "deriving" and "differentiating" mean quite different things. Could be `inderivative()` or `Zygote.isactive()`? Or since this is specific to reverse-mode, maybe `isreversemode()`, `haspullback()`?

**Comment by ChrisRackauckas** [comment]
This should definitely include tests, with #850 as an example.

**Comment by DhairyaLGandhi** [comment]
Added #850 as a test, as well as a second order derivative example.

How about `ispullback`?

**Comment by mcabbott** [comment]
It's a little weird that `Zygote.pullback(f, x)` is the function that *returns* what I'd call the pullback, the closure. What this function is testing is whether you are inside the former, not the latter. Which Zygote also calls the adjoint. 

`ispullback` sounds to me like it ought to tell you about its argument, like `iseven`. And `inpullback` sound like you are testing whether you are on the backward pass.

Maybe `withgrad()`, `withadjoint()` are possibilities?

**Comment by oxinabox** [comment]
this is another one like `ignore` that should be in ChainRules/ChainRulesCore.
but I won't block this over that

File: src/lib/utils.jl [hunk 1]
```diff
@@ -111,3 +111,28 @@ a real.
      1  0
 """
 hessian(f, x::AbstractArray) = forward_jacobian(x -> gradient(f, x)[1], x)[2]
+
+"""
+   isderiving()
```
Comment (DhairyaLGandhi) [inline src/lib/utils.jl:116]:
```suggestion
   isderiving()
   isderiving(x)
```

**Comment by DhairyaLGandhi** [comment]
Beyond the name, this shows that #850 can be had with this design, so we should be okay

**Comment by CarloLucibello** [comment]
So is this supposed to give no error?
```julia
  function f(x, y)
    @assert !Zygote.isderiving(x)
    @assert Zygote.isderiving(y)
    x + y
  end

gradient(y -> f(1, y), 1)
```

File: src/lib/utils.jl [hunk 2]
```diff
@@ -111,3 +111,29 @@ a real.
...
+   isderiving(x)
+
+Check whether the current function call is happening while taking the derivative.
+
+
+    julia> function f(x)
```
Comment (CarloLucibello) [inline src/lib/utils.jl:122]: this should be inside a ```julia-repl  block I think

File: src/lib/utils.jl [hunk 3]
```diff
@@ -111,3 +111,29 @@ a real.
      1  0
 """
 hessian(f, x::AbstractArray) = forward_jacobian(x -> gradient(f, x)[1], x)[2]
+
+"""
+   isderiving()
+   isderiving(x)
+
+Check whether the current function call is happening while taking the derivative.
```
Comment (CarloLucibello) [inline src/lib/utils.jl:119]: We should specify what is the argument `x`

**Comment by DhairyaLGandhi** [comment]
Pretty sure that errors.

File: src/lib/utils.jl (hunk 3 above)
Comment (DhairyaLGandhi) [reply src/lib/utils.jl:119]: This is common practice if the input can be generic - adding an `Any` adds no value

Comment (CarloLucibello) [reply src/lib/utils.jl:119]: I'm not talking about the type, I mean when one should use `isderiving()` and when `isderiving(x)`, the docstring only mentions a current function call without any reference to x

**Comment by DhairyaLGandhi** [comment]
bors try

I want to check buildkite doing its thing

**Comment by DhairyaLGandhi** [comment]
bump @ChrisRackauckas is we need to change something here

**Comment by DhairyaLGandhi** [comment]
Let have this in I think

bors r+

**Comment by CarloLucibello** [comment]
@DhairyaLGandhi why was this merged without review and a follow-up to the naming discussion?

**Comment by DhairyaLGandhi** [comment]
Please do suggest a better name if you some thoughts around it

**Comment by CarloLucibello** [comment]
also there a few unaddressed comments

**Comment by CarloLucibello** [comment]
Could you file a follow-up PR addressing them?

**Pull request merged**